USER_GUIDE.md

```diff
@@ -543,18 +569,48 @@ The program consists of a simple loop which counts down from `N` to 0.
 Each time the loop iterates once, it flips the value of the boolean in the
 second position of the input stack.
 
-Here is our burning question: is the final value of that boolean `True` or
-`False`?
+Here is our burning question: is the final value of that boolean, i.e., the
+value of the variable `$EVENNESS_RESULT`, `True` or `False`?
 
 Given just the information in the test above, it is impossible to know.
 With some careful thought, we may realize that if we additionally know whether
 the value `N` is even or odd, we can compute the final value of the program.
 That is, the result is `True` if `N` is even and `False` if `N` is odd.
+We can express this requirement as a postcondition:
+
+```
+# loop-parity.tzt - failing test with
+# - postcondition
+input { Stack_elt nat $N ; Stack_elt bool True } ; # N B
+code { INT ; DUP ;                                 # N N B
+       GT ;                                        # (N>0) N B
+       LOOP @I { DIP { NOT } ;                     # N ¬B
+                 PUSH int 1 ; SWAP ; SUB ;         # (N-1) ¬B
+                 DUP ;                             # (N-1) (N-1) ¬B
+                 GT                                # (N-1>0) (N-1) ¬B
+               } ;                                 # N B
+       DROP                                        # B
+     } ;
+output { Stack_elt bool $EVENNESS_RESULT } ;
+postcondition { # EVENNESS_RESULT = (N % 2) == 0
+                { PUSH nat 2 ; PUSH nat $N ;
+                  EDIV ; ASSERT_SOME ; CDR ;
+                  PUSH nat 0 ; CMPEQ ;
+                  PUSH bool $EVENNESS_RESULT ; CMPEQ } }
+```
 
-However, the K-Michelson test framework is not as clever as a programmer.
-Instead of using the shortcut we identified above, it will blindly execute the
-loop until termination. In other words, the test runner will loop infinitely
-by guessing different starting values of `N`.
+Unforunately, even after adding a postcondition, we still cannot complete the
+proof.
+The problem lies in the fact that the K-Michelson test framework verifies
+programs by performing symbolic execution, i.e., by executing the program all
+possible ways and checking that we get the result we want each time.
+In particular, this means the framework does *not know how many* times to
+execute a loop.
+Given any initial value of variable `N`, we can always pick a larger initial
+value.
+This means we will never be sure we have executed the loop *all possible ways*.
+In other words, the test runner will loop infinitely by guessing different
+initial values of `N`.
 
 How can we escape this endless cycle? The answer lies in a concept called a
 _loop invariant_ which allows us to summarize the unending behavior of a loop
```

Comment:
```suggestion
Unfortunately, even after adding a postcondition, we still cannot complete the
proof.
This is because the K-Michelson verifies
programs using symbolic execution, i.e., by executing the program all
possible ways and checking that we get the result we want each time.
In particular, this means the framework does *not know how many* times to
execute a loop.
Depending on the initial value of variable `N`, we may need to execute the loop body 100 times, or 500.
This means we will never be sure we have executed the loop *all possible ways*, and the runner will get stuck executing the program for larger and larger values of `N`.

How can we escape this endless cycle?
It seems clear that we shouldn't need to execute the loop in all possible ways --
we can argue by induction that the post-condition holds. 
We claim that the current value of $B$ equals `True` if the difference between initial value of $N$ and it's current value is even, otherwise it is `False`. When $N$'s initial value is `0` its clear that this since $B$ is initially `True`, and the body is not executed. When $N$ is greater than $0$, the body reduces the value of `$N` by `1` and flips `$B`. By applying the inductive hypothesis, our claim must hold.

This is called a _loop invariant_ . It allows us to summarize the properties of an loop we care about, so that we do not need to execute it to arbitrary depth.
```

Reply:
Shall we also use either `B` or `EVENNESS` consistently in both the variable names and in the comments in the program? Or perhaps use `EVEN`? Then use `EVEN_CURR` in the invariant, `EVEN_FINAL` in the post condition. Similarly, `N_INIT`, `N_CURR` and `N_FINAL`.